doc/design/logging.md

```diff
@@ -0,0 +1,396 @@
+# Logging
+
+## Summary and problem description
+
+Logging is a crucial part of a framework since it informs the developer and
+user of the framework about anomalies and helps with debugging by providing
+contextual information.
+
+Additionally, when integrated into a separate framework, the log messages should
+be forwarded in order to have a single logging infrastructure.
+
+The logging should also be performant and not allocate memory. Ideally it should
+be possible to disable it at compile time and let the compiler optimize it away.
+
+It should also be possible to alter the behavior of the logging via environment
+variables, e.g. the log level.
+
+## Terminology
+
+### Log levels and how to use them
+
+| LogLevel | Usage |
+|----------|-------|
+| FATAL    | For fatal and non-recoverable errors which essentially lead to termination of the program. |
+| ERROR    | For severe but recoverable errors. |
+| WARN     | For cases when something unintentional happens which is not considered a severe error. |
+| INFO     | For important status information a user should be aware of. |
+| DEBUG    | Anything that is helpful for debugging. |
+| TRACE    | Any comment could be replaced by logging call with log level trace. |
+
+The log levels `FATAL`, `ERROR` and `WARN` should not be used independently but
+in combination with the error handler.
+
+## Design
+
+### Considerations
+
+1. Similar to other iceoryx constructs, heap usage, exceptions and recursive
+   calls are forbidden. This also means that `std::string` and
+   `std::stringstream` cannot be used.
+
+2. Since the logger shall be used by the `iceoryx_hoofs` itself, the base
+   implementation can also not make use of functionality in e.g. `iox::cxx` or
+   `iox::posix`
+
+3. To have a minimal runtime overhead, string formatting shall only happen when
+   the log level is above the output threshold. This lazy evaluation shall also
+   apply to other potentially expensive operations like the execution of a
+   function call.
+
+4. Additionally, the minimal log level shall be configurable by a compile time
+   switch and everything below this log level should be compiled to no-ops. With
+   this, developer can make excessive use of the most verbose log level and
+   disable it by default. When debugging, this can be turned on and help to find
+   the root of the problem.
+
+5. There shall also be the possibility to alter the behavior of the logger via
+   environment variables. This can be functionality like setting the log level or
+   deciding whether the timestamps or file names shall be part of a log message.
+
+6. Since iceoryx is often used as transport layer in a higher level framework, it
+   shall be possible to replace the default logger at runtime and delegate the
+   log messages to the logger of such a framework. To prevent misuse, this can
+   only be done up to a specific time. Afterwards an error message shall be
```

Comment:
@elfenpiff @mossmaurice done

@elfenpiff we will definitely have to replace the logger at runtime. Let's assume someone want's to integrate iceoryx into an higher level framework and also want's to make use of the functionality of that framework, e.g. DDS. Since iceoryx will be lower level it would result in a cyclic dependency if this would only be possible at compile time, therefore it needs to be done at runtime.